cg/meta/workflow/nf_analysis.py

```diff
@@ -803,3 +835,68 @@ def get_cases_to_store(self) -> list[Case]:
+        """Return reference genome for a case.
+        If this information is not available in StatusDB, it defaults to hg38."""
+        reference_genome: set[str] = {
+            sample.reference_genome
+            for sample in self.status_db.get_samples_by_case_id(case_id=case_id)
+        }
+        if len(reference_genome) == 1:
+            return reference_genome.pop()
+        elif len(reference_genome) > 1:
+            raise CgError(
+                f"Samples linked to case {case_id} have different reference genome versions set"
+            )
+        else:
+            LOG.info(f"No reference genome specified, defaulting to {GenomeVersion.hg38}")
+
+    def get_gene_panel_genome_build(self, case_id: str) -> GenePanelGenomeBuild:
+        """Return build version of the gene panel for a case."""
+        reference_genome = self.get_reference_genome(case_id=case_id)
+        try:
+            return getattr(GenePanelGenomeBuild, reference_genome)
+        except AttributeError as error:
+            LOG.error(
+                f"Reference {reference_genome} has no associated genome build for panels: {error}"
+            )
+
+    def get_gene_panel(self, case_id: str, dry_run: bool = False) -> list[str]:
+        """Create and return the aggregated gene panel file."""
+        return self._get_gene_panel(
+            case_id=case_id,
+            genome_build=self.get_gene_panel_genome_build(case_id=case_id),
+            dry_run=dry_run,
+        )
+
+    def get_source_from_lims_by_sample_id(self, sample_id: str) -> str | None:
+        """Get the source from LIMS for a given sample ID."""
+        lims_sample: dict | None = self.get_lims_sample(sample_id=sample_id, silent_error=True)
+        return lims_sample.get("source", None) if lims_sample else None
+
+    def get_source_by_case(self, case_id: str) -> str:
+        """Get sources from LIMS for a given case.
+        Returns 'unknown' if sources cannot be retrieved.
+        Raises CgError if different sources are set for the samples linked to a case"""
+        sample_ids: Iterator[str] = self.status_db.get_sample_ids_by_case_id(case_id=case_id)
+        sources: set[str] = {
+            source
+            for sample_id in sample_ids
+            if (source := self.get_source_from_lims_by_sample_id(sample_id=sample_id)) is not None
+        }
+        LOG.info(f"sources are {sources}")
+        if not sources:
+            return "unknown"
+        if len(sources) == 1:
+            return sources.pop()
+        if len(sources) > 1:
+            raise CgError(
+                f"All samples in a case must be of the same tissue type. "
+                f"Different sources found for {case_id}: {sources}."
+            )
+
+    @staticmethod
+    def replace_non_alphanumeric(string: str, replace_by="_") -> str:
+        """Replace non-alphanumeric characters from a string."""
+        return re.sub(r"\W+", replace_by, string)
```

Comment:
This shouldn't be a method but a `TomteParameters` validator.

Reply:
And what do you think about using a more pythonic way instead of the regex?

strings have an `isalnum()` so we could iterate through the string and use a list comprehesion then to join them by the `replace_by` variable

Reply:
About your second suggestion, I think having a regex is easier to understand that and to follow than iterating and joining over values. So I will keep the regex

Reply:
👍